ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,623 @@
+      pruneUpToInclusive: Offset,
+      submissionId: Ref.SubmissionId,
+      pruneAllDivulgedContracts: Boolean,
+  ): CompletionStage[PruningResult] =
+    CompletableFuture.completedFuture(
+      PruningResult.ParticipantPruned
+    )
+
+  override def ledgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] =
+    Source.single(
+      LedgerInitialConditions(
+        ledgerId = ledgerId,
+        config = Configuration(
+          generation = 1L,
+          timeModel = LedgerTimeModel.reasonableDefault,
+          maxDeduplicationTime = java.time.Duration.ofSeconds(maxDedupSeconds.toLong),
+        ),
+        initialRecordTime = Timestamp.now(),
+      )
+    )
+
+  override def stateUpdates(
+      beginAfter: Option[Offset]
+  )(implicit loggingContext: LoggingContext): Source[(Offset, Update), NotUsed] = {
+    // TODO for PoC purposes:
+    //   This method may only be called once, either with `beginAfter` set or unset.
+    //   A second call will result in an error unless the server is restarted.
+    //   Bootstrapping the bridge from indexer persistence is supported.
+    synchronized {
+      if (stateUpdatesWasCalledAlready)
+        throw new IllegalStateException("not allowed to call this twice")
+      else stateUpdatesWasCalledAlready = true
+    }
+    logger.info("Indexer subscribed to state updates.")
+    beginAfter.foreach(offset =>
+      logger.warn(
+        s"Indexer subscribed from a specific offset $offset. This offset is not taking into consideration, and does not change the behavior of the ReadWriteServiceBridge. Only valid use case supported: service starting from an already ingested database, and indexer subscribes from exactly the ledger-end."
+      )
+    )
+
+    queueSource.merge(source)
+  }
+
+  val (queue: BoundedSourceQueue[Submission], queueSource: Source[(Offset, Update), NotUsed]) =
+    Source
+      .queue[Submission](submissionBufferSize)
+      .map { submission =>
+        val newOffsetIdx = offsetIdx.incrementAndGet()
+        val newOffset = toOffset(newOffsetIdx)
+        (newOffset, successMapper(submission, newOffsetIdx, participantId))
+      }
+      .preMaterialize()
+
+  logger.info(
+    s"BridgeLedgerFactory initialized. Configuration: [maxDedupSeconds: $maxDedupSeconds, submissionBufferSize: $submissionBufferSize]"
+  )
+
+  private def submit(submission: Submission): CompletionStage[SubmissionResult] =
+    toSubmissionResult(queue.offer(submission))
+
+  private def precomputeTransactionOutputs(
+      transaction: Transaction
+  ): Future[Either[SoxRejection, TransactionEffects]] = {
+    lazy val contractKeyInputs =
+      transaction.transaction.contractKeyInputs.left.map(invalidInputFromParticipant(transaction))
+    lazy val inputContracts = transaction.transaction.inputContracts
+    lazy val updatedContractKeys = transaction.transaction.updatedContractKeys
+    lazy val consumedContracts = transaction.transaction.consumedContracts
+
+    Timed.future(
```

Comment:
is this something which can take long time?
Not sure the Future and metering makes sense here